### PR TITLE
PyTorch C++ Codegen for differentiable functions

### PR DESCRIPTION
**Pull request by orausch**

### CODE REVIEW

File: daceml/autodiff/pytorch.py [hunk 1]
```diff
@@ -9,23 +9,24 @@
...
 
 
-def make_backward_function(model: ONNXModel,
-                           apply_strict=False
-                           ) -> Type[torch.autograd.Function]:
+def make_backward_function(
+    model: ONNXModel,
+    apply_strict=False
+) -> Tuple[dace.SDFG, dace.SDFG, BackwardResult, Dict[str, dt.Data]]:
     """ Convert an ONNXModel to a PyTorch differentiable function. This method should not be used on it's own.
```
Comment (tbennun) [inline daceml/autodiff/pytorch.py:24]:
```suggestion
    """ Convert an ONNXModel to a PyTorch differentiable function. This method should not be used on its own.
```

File: daceml/autodiff/pytorch.py [hunk 2]
```diff
@@ -9,23 +9,24 @@
...
         Instead use the ``backward=True`` parameter of :class:`daceml.pytorch.DaceModule`.
 
         :param model: the model to convert.
         :param apply_strict: whether to apply strict transformations before creating the backward pass.
-        :return: the PyTorch compatible :class:`torch.autograd.Function`.
+        :return: the PyTorch compatible :class:`torch.autograd.Function`. TODO Update
```
Comment (tbennun) [inline daceml/autodiff/pytorch.py:29]:
```suggestion
        :return: A 4-tuple of forward SDFG, backward SDFG, backward result, and input arrays for 
                 backward pass (as mapping of names to DaCe data descriptors).
```

File: daceml/autodiff/pytorch.py [hunk 3]
```diff
@@ -93,119 +94,33 @@ def make_backward_function(model: ONNXModel,
...
+        desc = backward_sdfg.arrays[bwd_name]
+        if isinstance(desc, dt.Scalar):
+            arr_name, arr_desc = backward_sdfg.add_array(bwd_name + "_array",
+                                                         [1],
+                                                         desc.dtype,
+                                                         transient=False,
+                                                         storage=desc.storage,
+                                                         find_new_name=True)
+            desc.transient = True
+            bwd_copy_state = backward_sdfg.add_state_before(backward_state,
```
Comment (tbennun) [inline daceml/autodiff/pytorch.py:116]: add_state_before or after?

File: daceml/pytorch/module_codegen.py [hunk 4]
```diff
@@ -23,8 +27,9 @@
 class CompiledTorchFunction:
     """ A tuple holding the context for an executable function """
     function: Callable  #: the torch callable function
-    compiled_sdfg: CompiledSDFG  #: the compiled SDFG holding the state
-    ptr: torch.Tensor  #: the state ptr argument to use when calling the function
+    compiled_sdfgs: List[
+        CompiledSDFG]  #: the compiled SDFGs holding their states
+    ptr: List[torch.Tensor]  #: the state ptrs to use when calling the function
```
Comment (tbennun) [inline daceml/pytorch/module_codegen.py:32]: what does "state ptr" mean? needs better doc

File: daceml/pytorch/module_codegen.py [hunk 5]
```diff
@@ -88,35 +96,64 @@ def initialize_outputs_code(module: 'daceml.pytorch.DaceModule',
...
-    # outputs
-    ptr_init_code += '\n    '.join(
-        f"{arglist[name].dtype.ctype} *{name}_ptr = {name}_.data_ptr<{arglist[name].dtype.ctype}>();"
+    ptr_init_code = "\n// setup input and output pointers\n"
+    for name in input_names:
+        ctype = arglist[name].dtype.ctype
+        if isinstance(arglist[name], data.Array) or dt.can_access(
+                dt.ScheduleType.GPU_Device, arglist[name].storage):
+            if name in guard_contiguous:
+                ptr_init_code += '\n' + f"Tensor {name} = {name}_.contiguous();"
```
Comment (tbennun) [inline daceml/pytorch/module_codegen.py:130]: extra underscore is still here.

File: daceml/pytorch/module_codegen.py [hunk 6]
```diff
@@ -88,35 +96,64 @@ def initialize_outputs_code(module: 'daceml.pytorch.DaceModule',
...
+            ptr_init_code += '\n' + f"{ctype} *{name}_ptr = {name}.data_ptr<{ctype}>();"
+
+        elif isinstance(arglist[name], data.Scalar):
+            if name in guard_contiguous:
+                ptr_init_code += '\n' + f"{ctype} {name}_ptr = {name}_.item().to<{ctype}>();"
```
Comment (tbennun) [inline daceml/pytorch/module_codegen.py:135]: same here, but the underscore in the definition above was removed. is this on purpose?

File: daceml/pytorch/module_codegen.py [hunk 7]
```diff
@@ -148,51 +185,206 @@ def item_to_cpp_literal(item) -> str:
...
         iterator = np.nditer(value.cpu().numpy(), order="C")
-        return f"{desc.dtype.ctype} {name}[{desc.total_size}] =" \
-               f" {{{', '.join(item_to_cpp_literal(e) for e in iterator)}}};"
+        gpu_copy_code = f"""
+        {desc.dtype.ctype} *{name}_ptr;
+        cudaMalloc(&{name}_ptr, ({sym2cpp(desc.total_size)}) * sizeof({desc.dtype.ctype}));
+        cudaMemcpyAsync({name}_ptr, {name}_ptr_cpu, ({sym2cpp(desc.total_size)}) * sizeof({desc.dtype.ctype}), cudaMemcpyHostToDevice, nullptr);
+        """
+        return f"""
+        {desc.dtype.ctype} {name}_ptr{'_cpu' if gpu_storage else ''}[{desc.total_size}] =
```
Comment (tbennun) [inline daceml/pytorch/module_codegen.py:197]:
```suggestion
        {desc.dtype.ctype} {name}_ptr{'_cpu' if gpu_storage else ''}[{sym2cpp(desc.total_size)}] =
```

File: daceml/pytorch/module_codegen.py [hunk 8]
```diff
@@ -148,51 +185,206 @@ def item_to_cpp_literal(item) -> str:
...
+            {forward_sdfg.name}Handle_t handle = reinterpret_cast<{forward_sdfg.name}Handle_t>(fwd_handle_ptr);
+
+
+            // call SDFG
+            __program_{forward_sdfg.name}(handle, {fwd_sdfg_call_arguments});
+
+            // save inputs/outputs for backward
+            {
+                f"ctx->save_for_backward({{{', '.join(f'{n}' for n in saved_io_for_backward)}}});" 
+                if saved_io_for_backward else ""
```
Comment (tbennun) [inline daceml/pytorch/module_codegen.py:311]: is this an f-string in an f-string?

File: daceml/pytorch/module_codegen.py [hunk 9]
```diff
@@ -217,42 +409,163 @@ def code_for_module(module: 'daceml.pytorch.DaceModule') -> str:
...
+}}
+"""
+
+
 def compile_and_get_function(module: 'daceml.pytorch.DaceModule',
                              dummy_inputs) -> CompiledTorchFunction:
     """ Get a torch callable for the module. This will compile the sdfg, compile a PyTorch C++ operator, register it
         with PyTorch and return the function that calls it.
 
+        This function handles both the forward and backward pass codegen.
```
Comment (tbennun) [inline daceml/pytorch/module_codegen.py:437]:
```suggestion
        This function handles code generation for both the forward and backward pass.
```

**Review by tbennun — CHANGES_REQUESTED** [review]
Only minor comments. Looks nice

File: daceml/pytorch/module_codegen.py (hunk 8 above)
Comment (orausch) [reply daceml/pytorch/module_codegen.py:311]: It's actually an f-string within an f-string within an f-string

File: daceml/pytorch/module_codegen.py (hunk 5 above)
Comment (orausch) [reply daceml/pytorch/module_codegen.py:130]: This is PyTorch convention: the externally passed tensor is named with an underscore, and once it is contiguous the underscore is removed

File: daceml/pytorch/module_codegen.py (hunk 6 above)
Comment (orausch) [reply daceml/pytorch/module_codegen.py:135]: See above

File: daceml/autodiff/pytorch.py (hunk 3 above)
Comment (orausch) [reply daceml/autodiff/pytorch.py]: Oh yeah, good catch. How did this work :confused:

**Comment by orausch** [comment]
@tbennun The new casts a bit annoying, but we need them because the pytorch binary only has symbols for `int64_t` (aka `long int`) not `long long int`. Since these should be the same I just reinterpret cast everything.

File: daceml/pytorch/module_codegen.py [hunk 10]
```diff
@@ -29,7 +29,19 @@ class CompiledTorchFunction:
     function: Callable  #: the torch callable function
     compiled_sdfgs: List[
         CompiledSDFG]  #: the compiled SDFGs holding their states
-    ptr: List[torch.Tensor]  #: the state ptrs to use when calling the function
+    #: the pointers to the initialized SDFG state handles. Must be passed as the first arguments to function.
+    ptr: List[torch.Tensor]
+
+
+_REPLACED_CTYPES = {dace.int64: "int64_t", dace.uint64: "int64_t"}
```
Comment (tbennun) [inline daceml/pytorch/module_codegen.py:36]:
```suggestion
_REPLACED_CTYPES = {dace.int64: "int64_t", dace.uint64: "uint64_t"}
```
?

**Review by tbennun — APPROVED** [review]

**Pull request merged**